enhancements/coreos-encrypted-by-default.md

```diff
@@ -0,0 +1,422 @@
+
+With the exclusion of the on-disk format, this work should lay the groundwork for an unattended encrypted boot that can be used by both FCOS and RHCOS.
+
+Ignition is the first boot agent for FCOS and RHCOS. FCOS discussed the topic of partition layouts, with the idea of having Ignition lay-down arbitrary root filesystem layouts. At this time, Openshift requires Ignition V2, while FCOS uses Ignition v3; these versions are incompatible. When Ignition supports the required depenendencies, a reviewion to this proposal will be forthcoming.
+
+## Terms
+* LUKS/LUKS2: the standard Linux encryption method. Encrypted filesystems stored in a container.
+* Master Key: the _actual_ encryption key used. The master key encrypted using a passphrase and is embedded in the LUKS meta-data on disk encrypted. This encryption key is unknown to users.
+* LUKS. Passphrase: any string or file that can be used to unlock the master key. Each passphrase is bound to a "key-slot."
+* LUKS Token: arbitrary JSON stored in the LUKS meta-data.
+* NBE: [Network Bound-Disk Encryption](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Security_Guide/sec-Using_Network-Bound_Disk_Encryption.html). A method of providing context-based disk encryption.
+* [Clevis](https://github.com/latchset/clevis): Linux CLI tools for policy-based encryption and decryption. Clevis has a Dracut module for automatically opening LUKS containers in an NBE context.
+* Clevis pin: a program called by Clevis to encrypt and decryption of secrets.
+* Tang: A server implementation that provides cryptographic binding services without the need for an escrow. Secrets are derived from what the client and the Tang server both have, reducing the attack surface.
+* KMS: Key Management System, such as Amazon KMS.
+* [TPM2](https://en.wikipedia.org/wiki/Trusted_Platform_Module): A secure cryptoprocessor for storing keys in hardware.
+* Dracut Module: program(s) that run early in boot before the on-disk Linux is booted.
+
+If you are unfamiliar with some of these terms [please see an excellent Youtube video on Clevis/Tang](https://youtu.be/2uLKvB8Z5D0).
+
+## Proposal
+
+To provide install-time _and_ rolling re-encryption, CoreOS boot images will:
+* Have the root filesystem housed in a LUKS container.
+* Use the 'null-cipher' for the initial LUKS container "encrpytion". This allows _any_ passphrase to unlocked the LUKS container.
+* The passphrase will be stored in a LUKS token via the Clevis plaintext pin.
+
+Note: the default boot image is _effectively *NOT* encrypted_.
+
+On first boot, a Dracut module will re-encrypt the LUKS container unless the kernel CLI parameter ('rd.coreos.no_encrpytion=1') is used.  The module will likely run after Ignition has completed. It will:
+  * Prepare a random passphrase,
```

Comment:
Would the admin have a way to record this?

Reply:
The random-passphrase would be used as part of the re-encryption step in key-slot one. After the LUKS binding for Clevis is done, it would be thrown away; it is needed only for the master-key generation and initial Clevis binding. 

An administrator would, however, be able to extract bound secret by reading the LUKS token and then manually decrypting it (e.g. `cryptsetup token export --token-id 0 | ... | clevis-decrypt`  The `...` is for a step I need look up for extracting the payload which is piped to clevis) .

Reply:
Hmm.  I'm trying to understand the flow of what happens by default on boot.  Later the doc says "if no TPM2 device, use the plaintext Clevis pin" - but what is that?

I guess another way to say this is - if I do a default install in AWS (no TPM), and I'm root on a machine, can I extract a key or passphrase to use for disaster recovery later, or no?

Reply:
I update the document: it's a proposed pin for the purpose of encoding and decoding the passphrase into a JOSE object. Clevis's LUKS binding operation will embed that JOSE object into a LUKS token. The pin's purpose it to store passphrase in a manner that Clevis understands for unattended boot.

Reply:
The plaintext Clevis pin is part of this proposal. Its purpose is to encode the plaintext passphrase into, and from, a JOSE object suitable for Clevis to use. This functionality allows for the instance to be encrypted from day one, and then the administrator can change to a different Clevis binding without having to re-encrypt the entire node.